src/libcore/mem.rs

```diff
@@ -22,15 +22,53 @@ use ptr;
 #[stable(feature = "rust1", since = "1.0.0")]
 pub use intrinsics::transmute;
 
-/// Moves a thing into the void.
+/// Leaks a value into the void, consuming ownership and never running its
+/// destructor.
 ///
-/// The forget function will take ownership of the provided value but neglect
-/// to run any required cleanup or memory management operations on it.
+/// This function will take ownership of its argument, but is distinct from the
+/// `mem::drop` function in that it **does not run the destructor**, leaking the
+/// value and any resources that it owns.
 ///
-/// This function is the unsafe version of the `drop` function because it does
-/// not run any destructors.
+/// # Safety
+///
+/// This function is not marked as `unsafe` as Rust does not guarantee that the
+/// `Drop` implementation for a value will always run. Note, however, that
+/// leaking resources such as memory or I/O objects is likely not desired, so
+/// this function is only recommended for specialized use cases.
+///
+/// The safety of this function implies that when writing `unsafe` code
+/// yourself, you cannot write a primitive that relies on a destructor running
+/// to preserve memory safety. Unsafe code must be resilient to destructors not
+/// running in all circumstances.
```

Comment:
I've pushed some minor updates to the wording here, but I will [echo @aturon](https://github.com/rust-lang/rfcs/pull/1066#issuecomment-100005585) in that explicitly specifying this kind of detail is likely best left to a future RFC with a more focused discussion.